hls_parser.c

```diff
@@ -0,0 +1,192 @@
+		"Couldn't find stream information.\n"
+	);
+	fmtCtx->flags |= AVFMT_FLAG_GENPTS;
+
+	// find video stream index
+	int videoindex = -1;
+	for (int i = 0; i < fmtCtx->nb_streams; i++) {
+		if (fmtCtx->streams[i]->codec->codec_type == AVMEDIA_TYPE_VIDEO) {
+			videoindex = i; break; } }
+	ASSERT(videoindex != -1, "Didn't find a video stream.\n");
+ 
+	// initialize inputCodecContext
+	AVStream 		*stream        = fmtCtx->streams[videoindex];
+	AVCodecContext	*inputCodecCtx = stream->codec;
+	AVCodec			*inputCodec    = avcodec_find_decoder(inputCodecCtx->codec_id);
+	ASSERT(inputCodec != NULL, "Codec not found.\n");
+	ASSERT(avcodec_open2(inputCodecCtx, inputCodec, NULL) >= 0, "Could not open codec.\n");
+
+	// initialize outputCodecContext
+	AVCodec        *outputCodec    = avcodec_find_encoder(AV_CODEC_ID_MJPEG);
+	ASSERT(outputCodec != NULL, "Codec not found.\n");
+	AVCodecContext *outputCodecCtx = avcodec_alloc_context3(outputCodec);
+	ASSERT(outputCodecCtx != NULL, "Could not open codec.\n");
+	outputCodecCtx->time_base = AV_TIME_BASE_Q;
+	outputCodecCtx->pix_fmt   = AV_PIX_FMT_YUVJ420P;
+	outputCodecCtx->width     = inputCodecCtx->width;
+	outputCodecCtx->height    = inputCodecCtx->height; 
+	outputCodecCtx->flags    |= AV_CODEC_FLAG_QSCALE;
+    outputCodecCtx->global_quality = FF_QP2LAMBDA * 1;
+	ASSERT(avcodec_open2(outputCodecCtx, outputCodec, NULL) >= 0, "Could not open codec.\n");
+	
+	// initialize frame and packet variables
+	AVPacket *packet   = av_packet_alloc();
+	AVFrame  *frame    = av_frame_alloc();
+	AVFrame  *frameYUV = av_frame_alloc();
+	unsigned char *out_buffer = (unsigned char *) av_malloc(
+		av_image_get_buffer_size(AV_PIX_FMT_YUV420P, inputCodecCtx->width, inputCodecCtx->height, 1)
+	);
+	av_image_fill_arrays(frameYUV->data, frameYUV->linesize, out_buffer,
+		AV_PIX_FMT_YUV420P, inputCodecCtx->width, inputCodecCtx->height, 1);
+	struct SwsContext *img_convert_ctx = sws_getContext(inputCodecCtx->width, inputCodecCtx->height, inputCodecCtx->pix_fmt, 
+		inputCodecCtx->width, inputCodecCtx->height, AV_PIX_FMT_YUV420P, SWS_BICUBIC, NULL, NULL, NULL); 
+
+	// output stream information
+	printf("--------------- File Information ----------------\n");
+	av_dump_format(fmtCtx, 0, filepath, 0);
+	printf("-------------------------------------------------\n");
+
+	// main loop
+	int got_picture = 0;
+	while (av_read_frame(fmtCtx, packet) >= 0) {
+		if(packet->stream_index == videoindex){
+			ASSERT(
+				avcodec_decode_video2(inputCodecCtx, frame, &got_picture, packet),
+				"Decode Error.\n"
+			);
+			if (got_picture) {
+				process(stream, inputCodecCtx, outputCodecCtx, img_convert_ctx, packet, frame, frameYUV);
+			}
+		}
+		av_free_packet(packet);
+	}
+	// flush all internal queues
+	while (avcodec_decode_video2(inputCodecCtx, frame, &got_picture, packet) >= 0 && got_picture) {
+		process(stream, inputCodecCtx, outputCodecCtx, img_convert_ctx, packet, frame, frameYUV);
+	}
+ 
+	// free up resources
+	av_frame_free(&frame);
```

Comment:
Should you also be freeing up `frameYUV`?

Reply:
`frameYUV` is being repeatedly used; freeing it up here would cause problems.

Reply:
Ah, you're right. I should free it up towards the end here.

Reply:
Updated in 6fef2c6